### ISSUE TITLE
Add explainer to warning notice when validation errors are displayed

### ISSUE THREAD

**Issue by westonruter** [Automattic/amp-wp#5304]
## Feature description

When Developer Tools are enabled, when a validation error occurs on the permalink for a post, a warning will be displayed in the editor:

> ![image](https://user-images.githubusercontent.com/134745/91778812-8c5a9080-eba8-11ea-85f8-c00ef650aec3.png)

Prior to 2.0, Developer Tools were not available for Reader mode. Since most sites are in Reader mode, many users are going to be seeing this warning notice for the first time. We need to explain to them why they are seeing this warning all of a sudden. They need to know that the validation errors were occurring before, but they just weren't being told about them.

I wrote a [support topic](https://wordpress.org/support/topic/why-you-are-now-seeing-amp-validation-errors-in-reader-mode-%e2%9a%a0%ef%b8%8f/) about this to preempt support topics being opened, but we could do something even more preemptive inside of WordPress. We can link to the relevant documentation page on amp-wp.org. If one isn't yet created, we should do so.

Additionally, we may want to provide a link to turn off Developer Tools in the user profile. Something like, “Don't want to see these warnings? Turn off <strong>AMP Developer Tools</strong> in your user profile” (where the bolded text links to the checkbox).

This is closely related to #4668, where we can mention here the specific themes and plugins that are causing validation errors. If there are validation errors from the theme, the warning should probably advise switching to Reader mode. Validation errors from plugins could advise to evaluate whether the removed invalid markup is needed, whether the plugin should be suppressed (with a link to the Plugin Suppression section). There can be prompts to reach out to the respective support forums for the themes and plugins.

Also related to #3821, as this information should be displayed in the sidebar.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

**Comment by amedina** [comment]
Related: add content to amp-wp.org with explainer, to be linked from a "Why I am seeing this" link.

**Comment by johnwatkins0** [comment]
I saw this notice on a client project just yesterday and experienced a small amount of confusion and nervousness. 

1. **Confusion:** "There are 2 issues from AMP validation which need review. The issues are not directly due to content here." The **where** of this is confusing. We should clarify to something like, "This post has two AMP validation issues, but they come from outside the post content. To review these issues, [go here]. To turn off these notices, [go here]."
2. **Nervousness:** Any time there is a link in a notice in the editor, I worried about what's going to happen when I click the link. I don't want the current tab to navigate to another page, potentially losing unsaved changes. The core UX around recovering autosaves is not great, so I avoid it in general. In other words, I think these links should have an icon indicating they open in a new tab. 

Should we go forward with minor copy tweaks along these lines for 2.0.1 and save the broader updates for 2.1?

**Comment by amedina** [comment]
These notices need to change. However, as they are right now they might be
confusing to some users, but the changes could also result in confusion. I
don't want to change them hastily. I would like to address this in the
context of the overall revamp of improving the "communication" aspects of
the validation tools, add the Document and block panels with a nice design
and precise messaging, and so on.

We will discuss soon (i.e. Roadmap planning).

**Comment by westonruter** [comment]
We can open the link in a new window, however.

**Comment by westonruter** [comment]
Please open a quick PR for opening the link on a new window. We'll include that in 2.0.1.

**Comment by amedina** [comment]
+1

**Comment by johnwatkins0** [comment]
In #5319 we made the change for 2.0.2 to open the link in a new tab. Should this be set to a later milestone now?

**Comment by westonruter** [comment]
You're right. Done.

**Comment by westonruter** [comment]
Update on this since #5589 was merged:

Removed and Unreviewed | Some Kept and Unreviewed | All Removed and Reviewed
---------------------------|----------------------------|-----------------------------
![image](https://user-images.githubusercontent.com/134745/103690881-362c9e00-4f4a-11eb-9ae9-1b1de0a74712.png) | ![image](https://user-images.githubusercontent.com/134745/103691101-7f7ced80-4f4a-11eb-86c3-10883503f970.png) | ![image](https://user-images.githubusercontent.com/134745/103691339-dc78a380-4f4a-11eb-9e73-17dcc5285890.png)

Notice there is only a notice in the case of there being kept markup. 

Contrast this with the validated URL screen:

Removed and Unreviewed | Some Kept and Unreviewed | All Removed and Reviewed
---------------------------|----------------------------|-----------------------------
![image](https://user-images.githubusercontent.com/134745/103691214-ae935f00-4f4a-11eb-9acd-cec360046077.png) | ![image](https://user-images.githubusercontent.com/134745/103691178-a20f0680-4f4a-11eb-9cca-85c8ed2c62fc.png) | ![image](https://user-images.githubusercontent.com/134745/103691252-ba7f2100-4f4a-11eb-872b-db854ba0b037.png)

I think we need to align these with the corresponding messages shown on the Validated URL screen.

Also, in the Status & Visibility panel of the Document sidebar, we need to indicate the AMP validation status with the toggle, especially when there is kept invalid markup. Currently it shows that AMP is enabled, but this is negated by the plugin sidebar icon which indicates it is not:

> ![image](https://user-images.githubusercontent.com/134745/103694447-d0431500-4f4f-11eb-8c9f-b9199f3e45a1.png)

Along with some notice, there should be a link or button that takes the user to the AMP plugin sidebar, similar in fact to how Yoast SEO has a button that does the same (also pictured here).

Lastly, any such AMP notice should also be displayed in the pre-publish dialog (similar to how Yoast does):

> ![image](https://user-images.githubusercontent.com/134745/103694592-0ed8cf80-4f50-11eb-9383-a0fe36dc8b93.png)

cc @jwold @delawski

**Comment by westonruter** [comment]
I just realized this is issue is not exactly what I thought it is, although it is related. I think I meant to comment this on the now-closed #3673 or #3821 (which were closed via #5589, which removed this global notice in favor of the plugin sidebar). In any case, this issue is still open and any AMP validation messaging that we show in the editor needs to be better tailored for non-technical users, so I think it makes sense to expand the scope.

**Comment by jwold** [comment]
Been playing with some concepts to address this, and leaning toward something like this:

**Some kept and unreviewed state:**

<img width="369" alt="Screen Shot 2021-01-06 at 1 55 26 PM" src="https://user-images.githubusercontent.com/3411173/103823435-eec80f80-5026-11eb-9f2a-9c5bfc4a6263.png">

**Comment by westonruter** [comment]
@jwold I think we can just put the "Invalid Markup Kept" info inside of the "AMP Disabled" panel. So in general I think we can iterate on this element here:

> ![image](https://user-images.githubusercontent.com/134745/103833825-bdefd680-5036-11eb-9ac8-e861dde73c91.png)

This would be the “AMP Status” box and it would have 3 possible states which correspond to the 3 states in the icon:

Icon | State
-----|------
![image](https://user-images.githubusercontent.com/134745/103833974-0effca80-5037-11eb-8bb1-10001ad44851.png) |  AMP Enabled, with either no validation issues or all markup being removed and reviewed.
![image](https://user-images.githubusercontent.com/134745/103833915-f394bf80-5036-11eb-8530-adf53be3e978.png) | AMP Enabled, but there are one or more instances of invalid markup that is unreviewed.
![image](https://user-images.githubusercontent.com/134745/103833954-03140880-5037-11eb-8b4c-ad7df51a5fd0.png) | AMP Disabled due to some invalid markup being kept on the page.

We can tailor the message for each state for non-technical end-users.

As for the "Show reviewed errors", I think it can be made less prominent. I know I raised a concern about a toggle appearing at the bottom of the list being problematic because of content shifting, but I think the reality is that this UI element makes better sense at the bottom since it should be giving minimal prominence. Clicking the link can cause the link to change to “Hide reviewed issues (x)”. This link would only be present if there any reviewed issues. If not, the link would be absent.

**Comment by jwold** [comment]
@westonruter I believe this captures all the comments above, and will be worth discussing on our next call.

<img width="609" alt="My Sketches - 2021-01-11 14 38 54" src="https://user-images.githubusercontent.com/3411173/104369948-085bd200-54d3-11eb-90a0-346aa264fef2.png">

My next step is to take this to a prototype stage, but I'd love any feedback first on it. Thanks!

cc @amedina and @delawski

**Comment by westonruter** [comment]
> I believe this captures all the comments above, and will be worth discussing on our next call.

Let's include your feedback from https://github.com/ampproject/amp-wp/pull/5741#issuecomment-763129207.

**Comment by jwold** [comment]
<img width="1116" alt="Screen Shot 2021-01-28 at 12 18 31 PM" src="https://user-images.githubusercontent.com/3411173/106193903-f9c41a80-6162-11eb-8495-c1446439b55b.png">

<img width="1195" alt="Screen Shot 2021-01-28 at 12 19 04 PM" src="https://user-images.githubusercontent.com/3411173/106193944-08123680-6163-11eb-97b2-e7b9c08bed74.png">

<img width="1154" alt="Screen Shot 2021-01-28 at 12 19 26 PM" src="https://user-images.githubusercontent.com/3411173/106193979-15c7bc00-6163-11eb-96e7-335811b02670.png">

@westonruter, @amedina, @delawski the latest revision of the work is ready for review. We chatted about this in our call today, and I'd love feedback. Thanks!

**Comment by westonruter** [comment]
@jwold sorry for the delay in reviewing these.

> ![image](https://user-images.githubusercontent.com/134745/107715113-6a028e00-6c83-11eb-98d7-3821edf20d05.png)

What again is the meaning behind the styling of the three errors listed here? The first one is kept but the last one is not, but they both have the orange border. I recall @schlessera having a good suggestion for the orange border, that it would only indicate the kept state so we wouldn't need the “kept” label necessarily. The “kept” label will be problematic because it will mash together with the error title, e.g. “Invalid inline script Kept”.

The four states, in order of frequency, are:

1. Unreviewed and removed (the default state)
2. Reviewed and removed (the normal resolution)
3. Unreviewed and kept (only happens by default for CSS)
4. Reviewed and kept (uncommon, as it causes AMP to be blocked for the page)

A final one in this context would be the partially-transparent error indicating the error was in a block that was removed, but that can be applied to any existing styling above.

I remember we had a discussion a couple weeks ago and deciding to diverge from how we've been trying to maintain the design language of unmoderated vs moderated comments, but now I forget. There's also the styling of suppressed plugins:

![image](https://user-images.githubusercontent.com/134745/107715749-cb772c80-6c84-11eb-835d-d0b20c34aef4.png)

**Comment by delawski** [comment]
@jwold @westonruter We may need to adjust the copy around the "Re-validate" message because of the way custom meta boxes are handled in the block editor. See https://github.com/ampproject/amp-wp/pull/5741#issuecomment-783508622.

As discussed, we may want to drop the message about post content change and leave the button alone.

**Comment by jwold** [comment]
Great feedback @delawski and @westonruter. A few things:

1. What if we combine the `re-validate` button with the `AMP status` message? (see design)
2. I was working on two things in parallel. What the design language should be for this update, and then what we'd want to for a future release. More in this direction: https://d.pr/i/KnngTd. However, I think we can pull parts of that in now. 
3. Not sure how we'd want to handle suppressed plugins. Any thoughts are welcome.

<img width="930" alt="Screen%20Shot%20on%202021-02-23%20at%2016-23-46" src="https://user-images.githubusercontent.com/3411173/108921419-8bed0080-75f3-11eb-81c5-406622124d83.png">

**Comment by delawski** [comment]
> 1. What if we combine the re-validate button with the AMP status message? (see design)

So if there are no custom meta boxes active, then the UI (the notifications area) would still look like this, correct?

<img width="1116" alt="Screen Shot 2021-01-28 at 12 18 31 PM" src="https://user-images.githubusercontent.com/3411173/106193903-f9c41a80-6162-11eb-8495-c1446439b55b.png">

Then, for the just discovered case when there are custom meta boxes active, we would use the new UI as you proposed:

<img width="930" alt="Screen%20Shot%20on%202021-02-23%20at%2016-23-46" src="https://user-images.githubusercontent.com/3411173/108921419-8bed0080-75f3-11eb-81c5-406622124d83.png">

I think it would work. Still, having an analogical UI but with different microcopy for both use cases would be better I guess.

> 2. I was working on two things in parallel. What the design language should be for this update, and then what we'd want to for a future release. More in this direction: https://d.pr/i/KnngTd. However, I think we can pull parts of that in now.

In general – yes, but I'd also say that it depends. Doing too invasive updates to the UI can easily become a rabbit hole. At the same time, we're still targeting the 2.1 release. So I guess it depends on which parts of the UI have you been thinking of.

**Comment by jwold** [comment]
@delawski good point. Let's separate the messages and keep them in different notifications. The wording will need to be massaged of course, thinking something like:

If `no` custom meta boxes:
> Page content has changed. Re-validate.

If `yes` custom meta boxes:
> Page content may have changed. Re-validate.

cc @westonruter

**Comment by westonruter** [comment]
> If `yes` custom meta boxes:
> 
> > Page content may have changed. Re-validate.

How about just “**Re-validate**” in this case? Is anything else needed? Or “Made changes? **Re-validate**”.

**Comment by westonruter** [comment]
@delawski @jwold Also, can this be designed in a way that content shifting is minimized? The loading spinner and validate button (as of #5741) both currently contribute to the UI shifting up and down, which is not ideal:

https://user-images.githubusercontent.com/134745/109264998-b1783680-77ba-11eb-8be7-9cbdbc005ef5.mov

**Comment by delawski** [comment]
> Also, can this be designed in a way that content shifting is minimized?

I was thinking about this issue, too.

I spoke about the loading state with @jwold lately and he suggested a pretty nice approach:

> <img width="625" alt="Screen Shot on 2021-02-25 at 12-07-47" src="https://user-images.githubusercontent.com/478735/109310228-8df9ce00-7844-11eb-86e2-e9f90ee3e1f4.png">
>
> [For the loading state, think of a situation] where a notification still shows, so it's more clear what's happening, and everything else disappears as it needs to. Also might require shifting the notification up to the top so it doesn't jump around.

@jwold Please chime in as I think we're going in a really good direction with that.

**Comment by delawski** [comment]
Another issue I've noticed while working on [#5741](https://github.com/ampproject/amp-wp/pull/5741#discussion_r583658462) is that we don't currently have any error handling.

If the REST endpoint responds with 4xx or 5xx, we're showing the loading spinner forever.

We should probably handle such errors with a proper message (sometimes a message is returned in the response – we could use it) and offer some support like a "Try again" button.

**Comment by jwold** [comment]
What if we did something like this? I think it might avoid content shifting altogether. 

<img width="743" alt="Screen Shot 2021-02-28 at 8 19 18 AM" src="https://user-images.githubusercontent.com/3411173/109425562-ab8e7b00-799d-11eb-92d9-b1b566f74d18.png">

**Comment by westonruter** [comment]
**QA Passed**

The scope of this issue evolved. The warning notice has been eliminated in favor of showing the error in the sidebar:

![image](https://user-images.githubusercontent.com/134745/115980748-de8b6000-a543-11eb-8b5b-a39258b6b57e.png)

And in the Post sidebar:

![image](https://user-images.githubusercontent.com/134745/115980756-f531b700-a543-11eb-994c-618a924d03bc.png)

And similarly in the pre-publish panel:

![image](https://user-images.githubusercontent.com/134745/115980753-ec40e580-a543-11eb-92b6-fce3ac147d9c.png)

When Developer Tools are turned off, the AMP Validation sidebar is not present:

![image](https://user-images.githubusercontent.com/134745/115980786-33c77180-a544-11eb-8258-a262d1006464.png)
